### DISCUSSION TITLE
Kafka bridge

### DISCUSSION THREAD

**Discussion by obschepit** [emqx/emqx#4601] [Q&A] [answered]
Hi!
Can i set key:value and headers when using Kafka bridge? Or only value (payload) is available?

**Comment by zmstone** [comment, marked as answer]
Hi @obschepit 
Key is supported in the same way as value.
e.g. if `"key": "${clientid}"` is provided, the Kafka message key will be the client ID.
 
Headers are not supported.

**Reply by zmstone** [reply]
will follow up with an update to the doc:
https://docs.emqx.io/en/enterprise/v4.2/bridge/bridge-kafka.html#configure-kafka-cluster

**Reply by obschepit** [reply]
Thanks.
Can i do this over Rule or Resources on GUI?

**Reply by zmstone** [reply]
Well, the API supports key, but key is not exposed for editing from the UI.
![image](https://user-images.githubusercontent.com/164324/115065780-e1a4a180-9eee-11eb-91a4-514aacfca9fe.png)

**Reply by zmstone** [reply]
will ping our frontend developers to have a look.

**Reply by zmstone** [reply]
Sorry. I was wrong.
When "Produce strategy" is set to `first_key_dispatch`, "Stragegy Key" can be set to clientId, username or (MQTT) topic.
![image](https://user-images.githubusercontent.com/164324/115103596-66290b80-9f53-11eb-970c-3d933a225187.png)
So, it's supported, but not as flexible

**Reply by zmstone** [reply]
confirm tested with this query: `SELECT payload.msg as msg, clientid FROM "t/#"`
and this message publish: `mosquitto_pub -h localhost -t 't/xyz' -m '{"msg":"hellox"}'`

I am working on a pull request to (to be released in enterprise edition 4.3.0):
1. Rename option "first_key_dispatch" to "key_dispatch" (which is a hash based Kafka partitioner);
2. "Strategy Key" option to be renamed to "Message Key";
3. Delete restriction hint about "Strategy Key" can only be used in conjunction with "fist_key_dispatch" (there is no such restriction, wrong hint message).

**Reply by obschepit** [reply]
OK. Thanks for the clarifications.
